Rewrites HandleSummary() doc

MattDodsonEnglish opened the pull request:
Besides sentence editing and content restructuring, this:
- Adds a small explanatory section
- Makes a progressive series of examples to show how it works
- Splits the reference into a separate section and describes the data model a bit more

Fixes: https://github.com/grafana/k6-docs/issues/679

MattDodsonEnglish commented:
By the way, I intentionally left some of the summarytrend options in the `textSummary` function undocumented. These options would get overwritten by script options, (I think), and it seems like using them would just cause unnecessary confusion.

MattDodsonEnglish commented on src/data/markdown/translated-guides/en/03 Results output/100 End-of-test/150-custom-summary.md:
```diff
@@ -1,35 +1,76 @@
+
+Fundamentally, `handleSummary()` is just a function that can access a data object.
+As such, you can transform the summary data into any text format: JSON, HTML, console, XML, and so on.
+You can pipe your custom summary to standard out or standard error, write it to a file, or send it to a remote server.
```
```suggestion
You can pipe your custom summary to [standard output or standard error](https://en.wikipedia.org/wiki/Standard_streams), write it to a file, or send it to a remote server.
```

MattDodsonEnglish commented on src/data/markdown/translated-guides/en/03 Results output/100 End-of-test/150-custom-summary.md:
```diff
@@ -41,7 +82,227 @@ The keys determine where k6 displays or saves the content:
-To get an idea how `data` would look in your specific test run, just add `return { 'raw-data.json': JSON.stringify(data)};` in your `handleSummary()` function and inspect the resulting `raw-data.json` file. Here's a very abridged example of how it might look:
+### Extract data properties
+
+This minimal `handleSummary()` extracts the `median` value for the `iteration_duration` metric and prints it on standard out:
```
```suggestion
This minimal `handleSummary()` extracts the `median` value for the `iteration_duration` metric and prints it to standard output:
```

MattDodsonEnglish commented on src/data/markdown/translated-guides/en/03 Results output/100 End-of-test/150-custom-summary.md:
```diff
@@ -1,35 +1,76 @@
-The keys determine where k6 displays or saves the content:
+The keys must be strings.
+They determine where k6 displays or saves the content:
 - `stdout` for [standard output](https://en.wikipedia.org/wiki/Standard_streams)
 - `stderr` for standard error,
 - any relative or absolute path to a file on the system (this operation overwrites existing files)
```
```suggestion
- `stdout` for standard output
- `stderr` for standard error,
- any relative or absolute path to a file on the system (this operation overwrites existing files)
```

MattDodsonEnglish commented on src/data/markdown/translated-guides/en/03 Results output/100 End-of-test/150-custom-summary.md:
```diff
@@ -1,35 +1,76 @@
+The value of a key can have a type of either `string` or [`ArrayBuffer`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer).
+
+You can return mutiple summary outputs in a script.
+As an example, this `return` statement sends a report to standard out and writes the `data` object to a JSON file.
```
```suggestion
As an example, this `return` statement sends a report to standard output and writes the `data` object to a JSON file.
```

ppcano commented on src/data/markdown/translated-guides/en/03 Results output/100 End-of-test/150-custom-summary.md:
```diff
@@ -1,35 +1,76 @@
+
+After your VU code runs, k6 aggregates your metrics into an object.
+The `handleSummary()` function takes this object as an argument (called `data` in all examples here).
+k6 calls `handleSummary()` at the end of the test run, even after [`teardown()`](/using-k6/test-life-cycle).
```
> k6 calls `handleSummary()` at the end of the test run, even after [`teardown()`](/using-k6/test-life-cycle).

I think this could be explained later.

ppcano commented on src/data/markdown/translated-guides/en/03 Results output/100 End-of-test/150-custom-summary.md:
```diff
@@ -1,35 +1,76 @@
-## Data format returned by `handleSummary()`
+## About the `handleSummary()` callback
+
+After your VU code runs, k6 aggregates your metrics into an object.
```
We mostly talk about test results and end-of-test summary. The word `object` does not sound familiar at this moment. 

One attempt:

> After k6 executes the load test, it aggregates the test results into a summary. The handleSummary() function receives this summary as an argument (called data in all examples here). 

> You can use handleSummary() to manipulate the test result summary as you want. To get an idea of what the data looks like, run this script and open the output file, summary.json.

mstoykov replied:
I am also against the wording of

> VU code runs

IMO this has always been confusing, but IMO in this case can just use "the load test has been executed" or some other variant of the "test was executed" instead of specyfing some code that has run and name it something,

MattDodsonEnglish replied:
Maybe I'm wrong, but to me this sentence is a key detail:

> Fundamentally, `handleSummary()` is just a function that can access a data object.

So, I'm mildly against removing the reference to `object` in other places because:
- It causes the doc to lose coherence (unless we remove all mentions of the data object).
- It's just less specific, calling it a "summary" doesn't describe the data as well as calling it a "summary object". And working with objects is a fundamental aspect of JavaScript programming, so I don't think it adds new vocabulary or unnecessary complication. Am i making an incorrect assumption here?


On the other hand, I agree with removing the reference to VU code and to callbacks, because these internal details don't really inform how a developer would code a `handleSummary()`.

mstoykov replied:
> So, I'm mildly against removing the reference to object in other places because:

I am also against not calling it `object` or `data` :thinking:. Object is bit vague in general ... so maybe `JavaScript object` just to make certain people understand that in this case we do mean a simpla javascript object.

> You can use handleSummary() to manipulate the test result summary as you want. 

IMO "manimulate" is the wrong word. You return the summary you want, you don't manipulate anything. In practice you can return w/e you want even something completely unrelated to the summary object provided.

ppcano commented on src/data/markdown/translated-guides/en/03 Results output/100 End-of-test/150-custom-summary.md:
```diff
@@ -41,7 +82,227 @@ The keys determine where k6 displays or saves the content:
 </CodeGroup>
 
-To get an idea how `data` would look in your specific test run, just add `return { 'raw-data.json': JSON.stringify(data)};` in your `handleSummary()` function and inspect the resulting `raw-data.json` file. Here's a very abridged example of how it might look:
+### Extract data properties
```
Suggest making it explicit that this is a (dummy) example.

### Example: extract data properties

??

ppcano commented on src/data/markdown/translated-guides/en/03 Results output/100 End-of-test/150-custom-summary.md:
```diff
@@ -41,7 +82,227 @@ The keys determine where k6 displays or saves the content:
+</CodeGroup>
+
+
+### Modify default output
```
Same here:

Example: modify default output

ppcano commented on src/data/markdown/translated-guides/en/03 Results output/100 End-of-test/150-custom-summary.md:
```diff
@@ -41,7 +82,227 @@ The keys determine where k6 displays or saves the content:
+
+### Custom reports to a remote server
+
+You can also send the generated reports to a remote server (over any protocol that k6 supports).
```
1 - Suggest the same:

- ### Example: custom reports to a remote server

2 - The example says `custom reports to a remote server` but it export to 3 outputs with different formats.

I suggest changing this example to showcase only the output to the remote server. 

If not, the title should be different.

ppcano reviewed (approved):
This PR significantly improves the article.  💯

I have suggested a few changes.

MattDodsonEnglish commented on src/data/markdown/translated-guides/en/03 Results output/100 End-of-test/150-custom-summary.md:
```diff
@@ -1,35 +1,76 @@
-
-Besides customizing the CLI summary, you can also transform the summary data into machine- or human-readable formats.
-This lets you create JS-helper functions that generate JSON, CSV, XML (JUnit/xUnit/etc.), HTML, etc. files from the summary data.
+With the `handleSummary()` callback, you can completely customize your end-of-test summary.
```
```suggestion
With the `handleSummary()` callback, you can completely customize your end-of-test summary.
```

Is there any reason to specify that it's a "callback" function? Or is this exposing too much internals? What function does it call back to?

If it's necessary to mention that it is a callback, maybe we could add an explanatory sentence at the end of the "About handleSummary()) section.

ppcano replied:
I am for Less is More. So, in this context, it could be unnecessary. 

> With `handleSummary()`, you can 


[Read more about callbacks](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Introducing#callbacks)

mstoykov replied:
I would argue it shouldn't be called callback function.

From the [MDN article](https://developer.mozilla.org/en-US/docs/Glossary/Callback_function) the definition is 
> a function passed into another function as an argument

We are missing the whole "passed into another function" so IMO we can drop it.

If we don't I would argue we should call `setup`, `teardown` and `default` callback functions as well ;)

ppcano replied:
I'd separate this convo in another thread. Here my thoughts:


1 - I think `default` would be different --> the default scenario function?

2 - But I agree `setup`, `teardown`, and `handleSummary` follow the same pattern.



3.1 - I have seen using `callback` for this case. But it is confusing.
3.2 - Another common term in JS land is [hook](https://mochajs.org/#hooks)
3.3 - To keep it simple - I am leaning toward using the term `global functions`?

MattDodsonEnglish replied:
Yes, this is beyond the scope of this PR, but I have two questions:

These functions are call backs to what other function? So it seems like these functions  run within a "super-function" that k6 automatically sets up when the script runs?

Is this detailed information about runtime useful to any users? If it is, we could probably add it to the Test life cycle document.

mstoykov replied:
> these functions run within a "super-function" that k6 automatically sets up when the script runs?

This won't be how I will explain it. K6 just happens to be calling this functions in a particular pattern that can be explained in the Test life cycle.

> I am leaning toward using the term global functions?

I am for "exported function"  as it (and the rest) need to be *exported* and a *function* so it seems like the best choice to me.

MattDodsonEnglish replied:
Sorry, just for my understanding, isn't there some kind of "superfunction" under the hood? I mean, _to what_ are these functions exported?

I know this is way too technical, and I beg forgiveness for my poor vocabulary, but it has confused me for a while. I think it's related to how I got so confused by "where" the data in `handlesummary(data)` came from.

ppcano replied:
I think the term "exported function" lacks some more context as any k6/JS module can export a function. 

Moving the convo to https://github.com/grafana/k6-docs/issues/845

mstoykov replied:
> Sorry, just for my understanding, isn't there some kind of "superfunction" under the hood? I mean, to what are these functions exported?

They are exported as [defined](https://developer.mozilla.org/en-US/docs/web/javascript/reference/statements/export) by the ecmascript specification. Which means that other JS modules can use them. 

k6 is not JS code it just uses the same mechanic to call them. 

I guess you can call it a superfunction (there is no such term, and it is not a [higher-order function](https://en.wikipedia.org/wiki/Higher-order_function)) it is just that k6 calls those functions with given parameters in a given ... pattern. They are [called](https://github.com/grafana/k6/blob/8a74171c8b43f48fb53078e0bcfa7851828124bc/js/runner.go#L321) from things that *are* functions within the go language :shrug:. 

But I am not really certain how this is relevant to anyone :shrug: 


>  got so confused by "where" the data in handlesummary(data) came from.

It comes from k6 providing it. 

In practice *almost* everything in computers comes from another program. When you run your program your OS provides it with a bunch of stuff in order to do it. In the same way we provide the info to the script when we execute it. It is just that OSes have some "standards" on how they will execute a program and we have some "standards" on how we will execute the script. And for the most parts most of the users do not care about *most* of those. 

In the cases of the functions we are talking about, you need to know that the first argument will be the data returned from setup or in the case of handleSummary the "summary data". Both of those are just things that when we designed the system decided will be useful. 

To be honest if we make this today - there is a good chance that some of those will just be things `k6/execution` will be able to return to you instead. As documenting that seems a lot easier and is definitely less confusing. Or at least we would have a discussion about it :shrug: 

Hope this helps you!

mstoykov commented on src/data/markdown/translated-guides/en/03 Results output/100 End-of-test/150-custom-summary.md:
```diff
@@ -1,35 +1,76 @@
+The `handleSummary()` function takes this object as an argument (called `data` in all examples here).
+k6 calls `handleSummary()` at the end of the test run, even after [`teardown()`](/using-k6/test-life-cycle).
+
+You can use `handleSummary()` to manipulate this object as you want.
+To get an idea of what the data looks like,
+run this script and open the output file, `summary.json`.
```
I do remember there were some differences betweeen this and the file you get from the [summary export](https://k6.io/docs/using-k6/k6-options/reference/#summary-export). 

But I guess we should say that somewhere and also specify you can skip this whole thing by using it. :thinking:

mstoykov reviewed (approved):
LGTM :clap:

MattDodsonEnglish commented on src/data/markdown/translated-guides/en/03 Results output/100 End-of-test/150-custom-summary.md:
```diff
@@ -1,35 +1,76 @@
+The `handleSummary()` function takes this object as an argument (called `data` in all examples here).
+k6 calls `handleSummary()` at the end of the test run, even after [`teardown()`](/using-k6/test-life-cycle).
+
+You can use `handleSummary()` to manipulate this object as you want.
```
```suggestion
You can use `handleSummary()` to create a custom summary or return the default summary object.
```

MattDodsonEnglish commented:
@ppcano and @mstoykov 
I thank you both for the great reviews, as always! :bow:  

I've removed all references to callbacks and VU code. I also removed the word manipulate.

I'm keeping object in, but I did qualify it as a _JavaScript object_.

With [db6b20d](https://github.com/grafana/k6-docs/pull/842/commits/db6b20dc70c5a63f1f16518ece7dc166bc9df4d4), I split up the examples and hopefully made them less trivial.

If it looks okay to you, Pepe, I think I'm happy to merge it.

ppcano commented:
@MattDodsonEnglish, I don't have further feedback. Thanks to you for these great improvements!
